tests/p4ctest.h

```diff
@@ -0,0 +1,118 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef P4CTEST_H
+#define P4CTEST_H
+
+#include <iostream>
+#include <vector>
+#include <string>
+
+/* Command handler context */
+struct p4c_cmdl_context {
+    /* number of command line arguments */
+    int argc;
+    /* array of command line arguments */
+    char **argv;
+    /* private context data defined by the API user */
+    void *pvt;
+};
+
+typedef void (*p4c_cmdl_handler)(struct p4c_cmdl_context *);
+
+struct p4c_cmdl_command {
+    const char *name;
+    const char *usage;
+    int min_args;
+    int max_args;
+    p4c_cmdl_handler handler;
+    enum { P4C_RO, P4C_RW } mode;  /* Does this command modify anythin? */
+};
+
+/* Overview
+ * ========
+ *
+ * P4C tests directory contains many small test programs. One down side
+ * of building them as individual programs is that they all have to be
+ * linked whenever a library function is modified.
+ *
+ * p4ctest is an attempt to improve the overall build time by linking
+ * all test programs into a single program, p4c-test. Regardless of
+ * the number of test programs, linking will be done only once to produce
+ * p4ctest.
+ *
+ * With p4ctest, each test programs now becomes a sub program of p4ctest.
+ * For example, 'mytest' program, can now be invoked as 'p4ctest mytest'.
+ *
+ * 'p4ctest --help' will list all test programs can be invoked.
+ *
+ * The 'Usage' section below documents how to add a new sub program
+ * to p4ctest using P4CTEST_REIGSTER macros.
+ */
+
+typedef void (*p4ctest_func)(int argc, char *argv[]);
+
+void p4ctest_register(const char *test_name, p4c_cmdl_handler f);
+
+void p4c_cmdl_run_command(struct p4c_cmdl_context *ctx,
+                          const std::vector<p4c_cmdl_command>& commands,
+                          bool read_only);
+
+/* Usage
+ * =====
+ *
+ * For each sub test program, its 'main' program should be named as
+ * '<test_name>_main()'.
+ *
+ * The 'main' programs should be registered with p4ctest as a sub program.
+ *    P4CTEST_REGISTER(name, function)
+ *
+ * 'name' will be name of the test program. It is expected as argv[1] when
+ * invoking with p4ctest.
+ *
+ * 'function' is the 'main' program mentioned above.
+ *
+ * Example:
+ * ----------
+ *
+ * Suppose the test program is called my-test.c
+ * ...
+ *
+ * static void
+ * my_test_main(int argc, char *argv[])
+ * {
+ *   ....
+ * }
+ *
+ * P4CTEST_REGISTER("my-test", my_test_main);
```

Comment:
How about REGISTER taking just one argument? The string should be generated from the test name; perhaps you can always append _main as well.